In [4]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding (fixed parameter name)
encoder = OneHotEncoder(sparse_output=False)   # ← this is the only change
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Build model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(4,)),
    tf.keras.layers.Dense(8, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

# Compile
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train
model.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)

# Evaluate
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data test: {acc:.4f}")

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Akurasi pada data test: 0.9667


## Tugas 2

- Ubah jumlah neuron hidden layer.
- Bandingkan akurasi dengan konfigurasi awal.

In [11]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding (fixed parameter name)
encoder = OneHotEncoder(sparse_output=False)   # ← this is the only change
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Build model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(16, activation='relu', input_shape=(4,)),
    tf.keras.layers.Dense(12, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

# Compile
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train
model.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)

# Evaluate
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data test: {acc:.4f}")

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Akurasi pada data test: 0.9667


Penjelasan
- Kapasitas sudah memadai Jika model awal sudah punya kapasitas lebih dari cukup untuk memisahkan kelas, menambahkan neuron ekstra tidak meningkatkan generalisasi — hanya menambah parameter yang tidak diperlukan.
- Overfitting vs Underfitting
Kedua model mungkin fit dengan sangat baik ke data training tanpa overfitting berlebih sehingga akurasi test tak berubah. Atau keduanya sedikit overfit tetapi performa test tetap sama.

Tugas 3

- Bandingkan Sigmoid vs ReLU pada dataset Iris.

- Catat perbedaan loss dan akurasi.

In [6]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding (fixed parameter name)
encoder = OneHotEncoder(sparse_output=False)   # ← this is the only change
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Build model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='sigmoid', input_shape=(4,)),
    tf.keras.layers.Dense(8, activation='sigmoid'),
    tf.keras.layers.Dense(3, activation='softmax')
])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
history_sig = model.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)

loss_sig, acc_sig = model.evaluate(X_test, y_test, verbose=0)
print("Sigmoid → Loss:", loss_sig, " Accuracy:", acc_sig)

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Sigmoid → Loss: 0.5614102482795715  Accuracy: 0.9333333373069763


In [8]:
#hidden layer (16,12)
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding (fixed parameter name)
encoder = OneHotEncoder(sparse_output=False)   # ← this is the only change
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Build model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(16, activation='sigmoid', input_shape=(4,)),
    tf.keras.layers.Dense(12, activation='sigmoid'),
    tf.keras.layers.Dense(3, activation='softmax')
])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
history_sig = model.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)

loss_sig, acc_sig = model.evaluate(X_test, y_test, verbose=0)
print("Sigmoid → Loss:", loss_sig, " Accuracy:", acc_sig)

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Sigmoid → Loss: 0.5548325777053833  Accuracy: 0.9666666388511658


Perbandingan
- ReLU hampir selalu unggul pada hidden layer
- ReLU tidak mengalami vanishing gradient
- Gradien besar → learning cepat
- Cocok untuk dataset kecil maupun besar
- Karena data Iris cukup mudah, ReLU menang stabil
- Itulah sebabnya akurasi ReLU = 0.9667 (96.67%) pada dua arsitektur

Sigmoid lebih lambat belajar → akurasi awal turun
- Pada arsitektur 10–8, hasil Sigmoid:
- 0.9333 atau 93.33%

Sigmoid membaik setelah neuron diperbesar
- Ketika neuron ditambah 16–12, akurasi Sigmoid meningkat menjadi:
- 0.9667 atau 96.67%
Yang berarti:

- Penambahan neuron memberi lebih banyak kapasitas
- Sigmoid butuh kapasitas lebih besar untuk belajar pola non-linear
- Lebih banyak neuron → gradien yang “menyelamatkan” training